### PR TITLE
Gnnexplainer

### PR DESCRIPTION
**Pull request by KounianhuaDu**
## Description
This is the implementation of gnn-explainer.  Five tasks on five synthetic datasets are conducted with a dummy gnn model. The visualization results are provided for reference.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by KounianhuaDu** [comment]
@zhjwy9343

**Comment by KounianhuaDu** [comment]
The pics are stored in my own repo. And the readme links to that repo. Is it appropriate？

File: examples/pytorch/gnn_explainer/explain_main.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+#-*- coding:utf-8 -*-
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/explain_main.py:1]: For Python 3, no need this line anymore. Pls remove it.

File: examples/pytorch/gnn_explainer/explain_main.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+        optim.zero_grad()
+        loss.backward()
+        optim.step()
+
+        #print("In epoch: {:03d}, Loss: {:.6f}, Pred: {}".format(epoch, loss.item(), model_predict[new_n_idx]))
+
+    # visualize the importance of edges
+    edge_weights = explainer.edge_mask.sigmoid().detach()
+    '''
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/explain_main.py:79]: Pls remove line 79 to 83. They are debugging codes.

File: examples/pytorch/gnn_explainer/explain_main.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                             'Will choose the first node in this class to explain')
+    parser.add_argument('--hop', type=int, default='2',
+                        help='The hop number of the computation sub-graph. For syn1 ans syn2, k=2. For syn3, syn4, and syn5, k=4.')
+    parser.add_argument('--epochs', type=int, default=200, help='The number of epochs.')
+    parser.add_argument('--lr', type=float, default=0.01, help='The learning rate.')
+    parser.add_argument('--wd', type=float, default=0.0, help='Weight decay.')
+    args = parser.parse_args()
+    print(args)
+
+    main(args)
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/explain_main.py:102]: Add a new line at the end

File: examples/pytorch/gnn_explainer/models.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+#-*- coding:utf-8 -*-
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/models.py:1]: Pls remove this line.

File: examples/pytorch/gnn_explainer/models.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    def forward(self, graph, n_feat, edge_weights=None):
+
+        h = self.in_layer(graph, n_feat, edge_weights)
+        h = F.relu(h)
+        h = self.hid_layer(graph, h, edge_weights)
+        h = F.relu(h)
+        h = self.out_layer(graph, h, edge_weights)
+
+        return h
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/models.py:59]: Add a new line at the end

File: examples/pytorch/gnn_explainer/train_main.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
+#-*- coding:utf-8 -*-
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/train_main.py:1]: Pls remove this line.

File: examples/pytorch/gnn_explainer/train_main.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    elif args.dataset == 'syn2':
+        g, labels, name = gen_syn2()
+    elif args.dataset == 'syn3':
+        g, labels, name = gen_syn3()
+    elif args.dataset == 'syn4':
+        g, labels, name = gen_syn4()
+    elif args.dataset == 'syn5':
+        g, labels, name = gen_syn5()
+    else:
+        raise ValueError()
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/train_main.py:32]: Please add some error message in the ValueError.

File: examples/pytorch/gnn_explainer/train_main.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    parser.add_argument('--feat_dim', type=int, default=10, help='The feature dimension.')
+    parser.add_argument('--hidden_dim', type=int, default=40, help='The hidden dimension.')
+    parser.add_argument('--epochs', type=int, default=500, help='The number of epochs.')
+    parser.add_argument('--lr', type=float, default=0.001, help='The learning rate.')
+    parser.add_argument('--wd', type=float, default=0.0, help='Weight decay.')
+
+    args = parser.parse_args()
+    print(args)
+
+    main(args)
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/train_main.py:86]: Add a new line at the end.

File: examples/pytorch/gnn_explainer/utils_graph.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
+#-*- coding:utf-8 -*-
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/utils_graph.py:1]: Pls remove this line.

File: examples/pytorch/gnn_explainer/utils_graph.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    ----------
+    graph:      DGLGraph, the full graph to extract from. This time, assume it is a homograph
+    seed_nodes:  Tensor, index of a node in the graph
+    hops:       Integer, the number of hops to extract
+
+    Returns
+    -------
+    sub_graph: DGLGraph, a sub graph
+    origin_nodes: List, list of node ids in the origin graph, sorted from small to large, whose order is the new id. e.g
+               [2, 51, 53, 79] means in the new sug_graph, whose node id is [0,1,2,3], the mapping is 2<>0, 51<>1, 53<>2,
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/utils_graph.py:31]: whose -> their new

File: examples/pytorch/gnn_explainer/utils_graph.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+                   "connectionstyle":"arc3,rad=0.1"}
+
+    nx.draw(g, pos, with_labels=True, node_color='b', **options)
+    if not (center_node is None):
+        nx.draw(g, pos, nodelist=center_node.tolist(), with_labels=True, node_color='r', **options)
+
+    plt.show()
+
+
+if __name__ == '__main__':
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/utils_graph.py:111]: Pls remove the main function.

File: examples/pytorch/gnn_explainer/NodeExplainerModule.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        6. Laplacian loss of the adj matrix.
+
+        In the PyG implementation, there are 5 types of losses:
+        1. The prediction loss between logits before and after node and edge masking;
+        2. Sum loss of edge mask weights;
+        3. Loss of edge mask entropy, which tries to put the mask value to either 0 or 1;
+        4. Sum loss of node feature mask weights;
+        5. Loss of node feature mask entropy, which tries to put the mask value to either 0 or 1;
+
+        In this implementation, will use the PyG's version, but set the aggregation function of weights
```
Comment (zhjwy9343) [inline examples/pytorch/gnn_explainer/NodeExplainerModule.py:151]: Please remove the ", but ..." and change to a period. I didn't set the aggregation function to be optional.

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Comment by zhjwy9343** [comment]
LGTM

**Pull request merged**